### DISCUSSION TITLE
Stitches future

### DISCUSSION THREAD

**Discussion by RobEasthope** [stitchesjs/stitches#1053] [Ideas]
Stitches has by far the best DX experience for writing styling code but there is a fair amount of concern regarding it's future especially regarding the changes coming in React 18.

I appreciate theat the team have been honest about the lack of official plans for the future and that the recent take over may have caused some change of plans.

Would it be possible to get a broad idea of the issues and possibilities for future development Stitches? Even if WorkOS doesn't have the dev time/resource to spare maybe the community could step up?

**Comment by hadihallak** [comment]
The future is static styles IMO — Streaming with concurrency mode makes runtime styles a pain to support with SSR.
you can read more about it [here](https://github.com/reactwg/react-18/discussions/110)  which is why i think that the best course of action for react 18 support is to turn Stitches into a static extraction library.

**Reply by RobEasthope** [reply]
Thanks @hadihallak, I had read that and it sounded alarm bells. Am I correct in thinking that some work was done on static extraction but it was abandoned a few months ago?

**Reply by hadihallak** [reply]
Yeah we did a prototype and the TLDR is that it's possible but the API would need some changes — mainly around how compounds and responsive variants interact with each others, as in, if you try to create the static compounds styles for multiple responsive variants they have an explosive nature and you would end up generating tons of styles that may or may not be used.

An example of that is lets say you had a config with 6 breakpoints and a compound variant that targets 3 variants (size, color, type) 

**Now to generate this file statically you would need to generate the cartesian product of all possible selectors for the above compound variants/breakpoints combo then inject the styles inside each of these new selectors:**
```
// pseudo selectors/medias
size @initial + color @initial +  type @initial
size @initial + color @bp1 +  type @initial
size @initial + color @ initial +  type @bp1
size @initial + color @ bp1 +  type @bp1
etc...
```

**You can see how this might get out of hand real quick and which is why i think we'd need to alter the API a little bit so that things the media config prop is separated into two different things:**
1- breakpoints (responsive variant will only be generated for these)
2- utility medias (for things like @hover or any utility media query that is only meant for usage inside the styles and isn't meant to create a new responsive variant)

the above separation would be a good start but an extra something that we can do is to alter how variants are created so that when you use `css({variants: {color: {}}})` it only creates normal non responsive variants and when you use `css({responsiveVariants: { size: {}}})` that is how we mark the variant as responsive and then it would work with the `Button({size: {"@bp1": 'xs'}})` syntax

I think these two main points are crucial for a static extraction version of stitches. 

-----

You might also wonder why wouldn't a static extraction version of stitches add only the used styles into that global sheet?
The answer would be is that it might be possible but probably very very very tricky because then the static extraction of stitches would need to act like a compiler in some way and know exactly which files created which components/variants and then only inject the styles that were used. 

Another very critical issue with trying to limit the injected styles to what's used by the app is that it will lead to issues when the used variants are not known build time like this example: `Button({variant: props.variant})`

there is simply no way for us to know what css this Button component will generate at build time simply because it depends on a runtime prop, and this is very common when you're building design systems and wrapping components.

**Reply by ruijdacd** [reply]
@hadihallak do you have roadmap in place with some, even rough estimations?

**Reply by hadihallak** [reply]
@ruijdacd for static extraction? no we don't have a roadmap as we haven't decided to support that yet.

we do have a milestone for the [2.0 release ](https://github.com/stitchesjs/stitches/milestone/1), but, we don't have a date set for it yet as this release involves a big rewrite to the internals to fix some very tricky bugs around injection order. 
I'm currently working on that and once we get the rewrite in place, the rest of tasks should flow swiftly

**Reply by fokkerone** [reply]
From modulz webpage

> We don’t have plans to introduce static extraction right now, but Stitches remains a great styling solution for dynamic web apps targeting desktop users primarily.

**Reply by hennessyevan** [reply]
I'm really interested to see what stylex looks like when(if) it arrives. In [that one demo](https://youtu.be/ur-sGzUWId4?t=306), they look at some common problems that arise with atomic-css (An initial goal of stitches was atomic css but it's really tricky and there are a lot of edge cases see stitches pre-v1). atomic-css would make static extraction more feasible by reducing the amount of css in general.

I think it would be great to watch stylex and either implement similar techniques or maybe stitches keeps its awesome DX as an abstraction over stylex or something 🤷🏻

**Reply by oleksify** [reply]
@hadihallak Are there any updates/estimates in regard to inject order fixes? I stumbled upon this issue now and having hard time figuring out the fix. Thanks!
